### PR TITLE
feat(query-engine-wasm): shrink gzipped Wasm size down to 1.3MB

### PR DESCRIPTION
**Pull request by jkomyno**
This PR contributes to https://github.com/prisma/team-orm/issues/584.

Before this PR, the gzipped `@prisma/query-engine-wasm` was 1.7MB.

- https://github.com/prisma/prisma-engines/pull/4530/commits/b5b8e2162f4b3286951220d1797e64cd25b5836c reduced the bundle size to 1.6MB:
  ```
  === Tarball Contents === 
  946B   README.md           
  355B   package.json        
  28.5kB query_engine_bg.js  
  4.3MB  query_engine_bg.wasm
  4.7kB  query_engine.d.ts   
  538B   query_engine.js     
  === Tarball Details === 
  name:          @prisma/query-engine-wasm 
  package size:  1.6 MB                                  
  unpacked size: 4.3 MB
  ```
- https://github.com/prisma/prisma-engines/pull/4530/commits/8e8c5202538aaffc1a4cd170125db106ab1141ca reduced the bundle size to 1.3MB:
  ```
  === Tarball Contents === 
  946B   README.md           
  355B   package.json        
  28.5kB query_engine_bg.js  
  3.3MB  query_engine_bg.wasm
  4.7kB  query_engine.d.ts   
  538B   query_engine.js     
  === Tarball Details (gzip) === 
  name:          @prisma/query-engine-wasm   
  package size:  1.3 MB                                  
  unpacked size: 3.3 MB
  ```

- https://github.com/prisma/prisma-engines/pull/4552/commits/398dd4a7c4336c0fd223ae64289879b5d8e2e9f3 kept the bundle size at 1.3MB, but it reveals that the unpacked size is ~0.2MB larger:
  ```
  === Tarball Contents === 
  946B   README.md           
  355B   package.json        
  28.5kB query_engine_bg.js  
  3.5MB  query_engine_bg.wasm
  4.7kB  query_engine.d.ts   
  538B   query_engine.js     
  === Tarball Details === 
  name:          @prisma/query-engine-wasm 
  package size:  1.3 MB                                  
  unpacked size: 3.5 MB
  ```
/integration

---

## Miscellaneous

Passing the `--reference-types` option to `wasm-pack build`, or the `--remove-non-js-ops` option to `wasm-opt` results in runtime errors when running the example. Reference types in particular would be good to have, as they would likely improve the CPU performance of `@prisma/query-engine-wasm`.

### CODE REVIEW

File: query-engine/query-engine-wasm/Cargo.toml [hunk 1]
```diff
@@ -46,3 +46,15 @@ tracing-subscriber = { version = "0.3" }
...
+wasm-opt = [
+    "-Os",                                # focus on code size
+    "--vacuum",                           # removes obviously unneeded code
+    "--duplicate-function-elimination",   # removes duplicate functions 
+    "--duplicate-import-elimination",     # removes duplicate imports
+    "--optimize-for-js",                  # early optimize of the instruction combinations for js
+    "--gsi",                              # globally optimize struct values
+    "--directize",                        # turns indirect calls into direct ones
+    "--cfp",                              # propagate constant struct field values
+]
```
Comment (jkomyno) [inline query-engine/query-engine-wasm/Cargo.toml:60]: TODO: extract the "optimize" flags, and place them into a new draft PR for "sometimes later in the future". We agreed on focusing on size first, performance later.

File: query-engine/query-engine-wasm/src/wasm/engine.rs [hunk 2]
```diff
@@ -149,7 +149,8 @@ impl QueryEngine {
         let env = stringify_env_values(env)?; // we cannot trust anything JS sends us from process.env
         let overrides: Vec<(_, _)> = datasource_overrides.into_iter().collect();
 
-        let mut schema = psl::validate(datamodel.into());
+        // Note: if we used `psl::validate`, we'd
```
Comment (jkomyno) [inline query-engine/query-engine-wasm/src/wasm/engine.rs:152]: TODO: add missing part of the comment

Comment (jkomyno) [inline query-engine/query-engine-wasm/src/wasm/engine.rs:152]:
```suggestion
        // Note: if we used `psl::validate`, we'd add ~1MB to the Wasm artifact (before gzip).
```

File: query-engine/query-engine-wasm/Cargo.toml [hunk 3]
```diff
@@ -46,3 +46,15 @@ tracing-subscriber = { version = "0.3" }
...
+[package.metadata.wasm-pack.profile.release]
+wasm-opt = [
+    "-Os",                                # focus on code size
+    "--vacuum",                           # removes obviously unneeded code
+    "--duplicate-function-elimination",   # removes duplicate functions 
+    "--duplicate-import-elimination",     # removes duplicate imports
+    "--optimize-for-js",                  # early optimize of the instruction combinations for js
+    "--gsi",                              # globally optimize struct values
+    "--directize",                        # turns indirect calls into direct ones
+    "--cfp",                              # propagate constant struct field values
```
Comment (jkomyno) [inline query-engine/query-engine-wasm/Cargo.toml:59]:
```suggestion
    "--vacuum",                           # removes obviously unneeded code
    "--duplicate-function-elimination",   # removes duplicate functions 
    "--duplicate-import-elimination",     # removes duplicate imports
```

**Review by SevInf — APPROVED** [review]

**Pull request merged**